File: src/root/doc/scripts/request_with_payload.py
#!/usr/bin/env python

import requests
import hashlib
import time
import sys


# return response
# throw exceptions in case of errors 
def putData(url, data, login, password):
    # calculate token & full URL
    identity = hashlib.sha1((login + password).encode('utf-8')).hexdigest()
    gbTime = str(int(time.time()))
    token = hashlib.sha1((url + identity + gbTime).encode('utf-8')).hexdigest()
    request = url + '&gbLogin=' + login + '&gbTime=' + gbTime + '&gbToken=' + token
    # send request & parse response
    res = requests.put(request, data=data)
    response = res.text
    # check status
    if res.status_code != 200:
        raise  Exception("Error for PUT requests: " + response)
    return response


# return response
# throw exceptions in case of errors 
def postData(url, data):
    # send request & parse response
    res = requests.post(url, data=data)
    response = res.text
    # check status
    if res.status_code != 200:
        raise  Exception("Error for POST requests: " + response)
    return response


if len(sys.argv) != 3 and len(sys.argv) != 5:
    sys.stderr.write("Incorrect parameters!\n")
    print("This program sends POST or PUT request with payload and print the response.")
    print("Parameters for POST request:  URL  file_with_payload")
    print("Parameters for PUT  request:  URL  file_with_payload  login  password")
    sys.exit()

url      = sys.argv[1]
file     = sys.argv[2]
if len(sys.argv) > 3:
    login    = sys.argv[3]
    password = sys.argv[4]
else:
    login = password = None

data = open(file).read()
if not url.startswith(("http://","https://")):
    url = "http://" + url

if login is None or password is None:
    response = postData(url, data)
else:
    response = putData(url, data, login, password)

print response

File: src/root/doc/scripts/request_with_payload.rb
#!/usr/bin/env ruby

require 'net/http'
require 'digest/sha1'


# return response
# throw exceptions in case of errors 
def putData(url, data, login, password)
    # calculate token & full URL
    identity = Digest::SHA1.hexdigest("#{login}#{password}")
    gbTime = Time.now.to_i.to_s
    token = Digest::SHA1.hexdigest("#{url}#{identity}#{gbTime}")
    request = "#{url}&gbLogin=#{login}&gbTime=#{gbTime}&gbToken=#{token}"
    # send request & parse response
    http = Net::HTTP.new(URI(url).host)
    http.read_timeout = 1200 # seconds
    req = Net::HTTP::Put.new("#{request}")
    req.body = data
    res = http.request(req)
    response = res.body
    # check status
    if not res.is_a? Net::HTTPSuccess
        raise "Error for PUT requests: #{response}"
    end
    return response
end


# return response
# throw exceptions in case of errors 
def postData(url, data)
    # send request & parse response
    http = Net::HTTP.new(URI(url).host)
    http.read_timeout = 1200 # seconds
    req = Net::HTTP::Post.new("#{url}")
    req.body = data
    res = http.request(req)
    response = res.body
    # check status
    if not res.is_a? Net::HTTPSuccess
        raise "Error for POST requests: #{response}"
    end
    return response
end


if ARGV.size != 2 and ARGV.size != 4
    STDERR.puts "Incorrect parameters!"
    puts "This program sends POST or PUT request with payload and print the response."
    puts "Parameters for POST request:  URL  file_with_payload"
    puts "Parameters for PUT  request:  URL  file_with_payload  login  password"
    exit 1
end

url=ARGV[0]
file=ARGV[1]
login=ARGV[2]
password=ARGV[3]

data = File.open(file).read
url = "http://" + url if (url !~ /^http:\/\// and url !~ /^https:\/\//)
response = nil

if login.nil? or password.nil?
  response = postData(url, data)
else
  response = putData(url, data, login, password)
end

puts "#{response}"
